alg/los.cpp

```diff
@@ -205,19 +205,113 @@ bool GDALIsLineOfSightVisible(const GDALRasterBandH hBand, const int xA,
         return true;
     }
 
-    // TODO if both X's or Y's are the same, it could be optimized for vertical/horizontal lines.
+    // Lambda for Linear interpolate like C++20 std::lerp.
+    auto lerp = [](const double a, const double b, const double t)
+    { return a + t * (b - a); };
+
+    // Lambda for getting Z test height given y input along the LOS line.
+    // Only to be used for vertical line checks.
+    auto GetZValueFromY = [&](const int y) -> double
+    {
+        // A ratio of 0.0 corresponds to being at yA.
+        const auto ratio =
+            static_cast<double>(y - yA) / static_cast<double>(yB - yA);
+        return lerp(zA, zB, ratio);
+    };
+
+    // Lambda for getting Z test height given x input along the LOS line.
+    // Only to be used for horizontal line checks.
+    auto GetZValueFromX = [&](const int x) -> double
+    {
+        // A ratio of 0.0 corresponds to being at xA.
+        const auto ratio =
+            static_cast<double>(x - xA) / static_cast<double>(xB - xA);
+        return lerp(zA, zB, ratio);
+    };
+
+    // Lambda for checking path safety of a vertical line.
+    // Returns true if the path has clear LOS.
+    auto CheckVerticalLine = [&]() -> bool
+    {
+        if (xA < xB)
+        {
+            for (int x = xA; x <= xB; ++x)
+            {
+                const auto zTest = GetZValueFromX(x);
+                if (!IsAboveTerrain(hBand, x, yA, zTest))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+        if (xA > xB)
+        {
+            for (int x = xA; x >= xB; --x)
+            {
+                const auto zTest = GetZValueFromX(x);
+                if (!IsAboveTerrain(hBand, x, yA, zTest))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // If we get here, it's a coding error.
+        return false;
```

Comment:
Could we also add an assertion error here somehow?

Reply:
I feel it is better to change the if (xA > xB) to a else to avoid code that will never be reached. You may add a CPLAssert(xA != xB) at the beginning of the lambda to make the contract clear.

Reply:
Good good. I added asserts that express the intent of the lamba much better than a comment.

The unit tests need more coverage on these - I found a copy-paste bug in these lambdas that wasn't caught.
I'll add more comprehensive testing once I get the other feature with reporting the LOS violation location.